apps/google-analytics-4/src/components/FormControlServiceAccountKey.tsx

```diff
@@ -0,0 +1,232 @@
+import React, { useCallback, useEffect, useState } from 'react';
+import {
+  Collapse,
+  Flex,
+  FormControl,
+  Note,
+  Paragraph,
+  Text,
+  Textarea,
+  TextLink,
+} from '@contentful/f36-components';
+import {
+  CheckCircleIcon,
+  ChevronDownIcon,
+  ChevronRightIcon,
+  ExternalLinkTrimmedIcon,
+} from '@contentful/f36-icons';
+import { css } from 'emotion';
+
+import tokens from '@contentful/f36-tokens';
+import InstalledServiceAccountKey from './InstalledServiceAccountKey';
+import {
+  convertServiceAccountKeyToServiceAccountKeyId,
+  convertKeyFileToServiceAccountKey,
+  AssertionError,
+} from '../utils/serviceAccountKey';
+import type { ServiceAccountKey, ServiceAccountKeyId } from '../types';
+
+interface FormControlState {
+  isInvalid: boolean;
+  errorMessage: string;
+}
+
+interface FormControlServideAccountKeyFileProps {
+  setServiceAccountKey: (serviceAccountKey: ServiceAccountKey | null) => void;
+  setServiceAccountKeyId: (serviceAccountKey: ServiceAccountKeyId | null) => void;
+  setIsInvalid: (isInvalid: boolean) => void;
+  currentServiceAccountKeyId: ServiceAccountKeyId | null;
+  className?: string;
+}
+
+const styles = {
+  credentialsInput: css({
+    fontFamily: 'monospace',
+  }),
+  credentialsNote: css({
+    marginBottom: tokens.spacingM,
+  }),
+  credentialsExplanation: css({
+    marginTop: tokens.spacingL,
+  }),
+};
+
+const placeholderText = `{
+  "type": "service_account",
+  ...
+}`;
+
+const FormControlServiceAccountKey = ({
```

Comment:
I am not really happy how state is managed in this component atm. There are four states in the parent and three states in here. Most of them are derived from `serviceAccountKeyFile`. At the moment they are kept in sync with `useEffect`s. To me, this looks very risky and easy to break if we introduce more in the future. I'd suggest pulling `serviceAccountKeyFile` in the parent component and derive all other values from that. Derived values can then use `useMemo` instead of manually updating them via `useEffect`. That should improve readability and it's easier to follow where the data comes from and what happens when you change it.

E.g. it's quite straight forward to derive the key and id from the file. Also whether it's valid or not can just be done with a `useMemo`.

Reply:
Okay, this was great advice. I moved all the state into the parent and refactored to pass event handlers down to the child. This had the nice effect of removing all the `useEffect`s and weird manipulation of state. Thanks for this helpful feedback!